### PR TITLE
fix: ensure the EdDSA key generation produces a 32 bytes public key

### LINKED ISSUE #4445: Flaky Test: CryptoConverterTest
# Bug Report

## Describe the Bug

CryptoConverterTest.createVerifierFor_edDsaKey_bouncyCastleProvider() is flaky. It has a very slight chance to fail.

## Steps to Reproduce

I cannot reproduce it after numerous attempt. I am sure it is not caused by cache as I already cleaned it. I am also sure it was not my code that make it fail as it later passed without any modification to my code, and that the stacktrace shows it is irrelevant to `QueryValidatorFactory`.

## Context Information

I ran `./gradlew test` in wsl2.

- Used version : the newest.
- OS: wsl2
- ...

## Detailed Description
```
java.lang.IllegalArgumentException: Given public key's length is not 32.
	at com.google.crypto.tink.subtle.Ed25519Verify.<init>(Ed25519Verify.java:98)
	at com.google.crypto.tink.subtle.Ed25519Verify.<init>(Ed25519Verify.java:83)
	at com.nimbusds.jose.crypto.Ed25519Verifier.<init>(Ed25519Verifier.java:107)
	at com.nimbusds.jose.crypto.Ed25519Verifier.<init>(Ed25519Verifier.java:78)
	at org.eclipse.edc.security.token.jwt.CryptoConverter.createEdDsaVerifier(CryptoConverter.java:382)
	at org.eclipse.edc.security.token.jwt.CryptoConverter.createVerifierFor(CryptoConverter.java:143)
	at org.eclipse.edc.security.token.jwt.CryptoConverterTest.createVerifierFor_edDsaKey_bouncyCastleProvider(CryptoConverterTest.java:139)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:766)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$8(TestMethodTestDescriptor.java:217)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:124)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:94)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:63)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:92)
	at jdk.proxy1/jdk.proxy1.$Proxy4.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:198)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:130)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:101)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:61)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:122)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

## Possible Implementation

_You already know the root cause of the erroneous state and how to fix it? Feel free to share your thoughts._

### PR DESCRIPTION
**Pull request by nikschul**
## What this PR changes/adds

This PR fixes an edge case in the Ed25519 public key encoding process, specifically within the encodeX (EdECPoint point) method. It introduces a normalization step that guarantees the extracted Y-coordinate is strictly 32 bytes long before we reverse the array, apply the isXOdd parity bit, and encode it to Base64URL.

## Why it does that

Previously, we relied directly on point.getY().toByteArray(). Because BigInteger dynamically sizes its output based on the mathematical value of the number, it introduces two unpredictable behaviors for cryptography:

- Dropped leading zeros: If the Y-coordinate happens to be small enough, BigInteger drops the leading zero byte, returning a 31-byte array. This resulted in truncated 42-character Base64URL strings (instead of the required 43 characters).

- Sign extension bytes: If the highest bit of a 32-byte number is 1, BigInteger prepends an extra `0x00` byte to keep the number positive, returning a 33-byte array.

Because the array length was dynamic, our isXOdd mask (bytes[bytes.length - 1] ^= mask) was occasionally being applied to the wrong byte index. Forcing a strict 32-byte array ensures standard-compliant Ed25519 JWKs and guarantees the parity bit is always applied to the correct index.

## Further notes

- Added a normalization block to handle stripping extra sign bytes or padding missing zero bytes.

- The isXOdd mask index is now safely hardcoded to bytes[31] since the array size is guaranteed.

- Two tests are added, to test for both edge cases with 31 byte and 33 byte keys

## Linked Issue(s)

Closes #4445
Closes [Identity Hub Issue #913](https://github.com/eclipse-edc/IdentityHub/issues/913)

### CODE REVIEW

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java [hunk 1]
```diff
@@ -459,12 +460,23 @@ private static Base64URL encodeD(EdECPrivateKey edKey) {
      */
     @NotNull
     private static Base64URL encodeX(EdECPoint point) {
-        var bytes = reverseArray(point.getY().toByteArray());
+        byte[] ybytes = point.getY().toByteArray(); // get the raw bytes from BigInteger
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: minor nit: `yBytes`

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java [hunk 2]
```diff
@@ -459,12 +460,23 @@ private static Base64URL encodeD(EdECPrivateKey edKey) {
...
+
+        byte[] fixedBytes = new byte[32]; // normalize to exactly 32 bytes (handling dropped zeros or extra sign bytes)
+        if (ybytes.length == 33 && ybytes[0] == 0) {
+            System.arraycopy(ybytes, 1, fixedBytes, 0, 32); // strip the extra leading zero added by BigInteger for sign
+        } else if (ybytes.length <= 32) {
+            System.arraycopy(ybytes, 0, fixedBytes, 32 - ybytes.length, ybytes.length); // pad with leading zeros if BigInteger dropped them
+        } else {
+            throw new IllegalArgumentException("Y-coordinate is too large for Ed25519");
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: not covered by tests

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 3]
```diff
@@ -459,4 +465,59 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
...
+
+    @Test
+    void createJwk_whenAsn1BytesAreCorrupted_throwsWrappedException() {
+        // 1. Mock an EdECPublicKey
+        var mockCorruptedKey = mock(EdECPublicKey.class);
+        when(mockCorruptedKey.getAlgorithm()).thenReturn(CryptoConverter.ALGORITHM_ED25519);
+        when(mockCorruptedKey.getParams()).thenReturn(NamedParameterSpec.ED25519);
+
+        // 2. Simulate garbage bytes that are not a valid ASN.1 structure
+        byte[] garbageBytes = new byte[] { 0x00, 0x01, 0x02, 0x03, 0x04 };
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]:
```suggestion
        var garbageBytes = new byte[] { 0x00, 0x01, 0x02, 0x03, 0x04 };
```

**Comment by nikschul** [comment]
> the code is now more than 2 years old, maybe Nimbus now has a way to convert JDK types (`PrivateKey` and `PublicKey`) into Nimbus types? @nikschul could you investigate this?

as far as I can see, there is no `java.security.interfaces.EdECPublicKey` constructor to the Nimbus `OctetKeyPair`

> the code is now more than 2 years old, maybe Nimbus now has a way to convert JDK types (`PrivateKey` and `PublicKey`) into Nimbus types? @nikschul could you investigate this?

@paullatzelsperger
Even in the newest version of Nimbus JOSE+JWT(10.8) there is no constructor for an `OctetKeyPair` from a `java.security.interfaces.EdECPublicKey` https://javadoc.io/doc/com.nimbusds/nimbus-jose-jwt/latest/com/nimbusds/jose/jwk/OctetKeyPair.html

This is likely because converting an EdECPublicKey to the raw bytes required by the JWK OKP format is non-trivial: the JDK represents the public key point's Y coordinate as a BigInteger, which is big-endian and strips leading zeros, whereas Ed25519 requires a fixed 32-byte little-endian encoding. Thus, a naive conversion would silently produce incorrect key material in some cases.

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java [hunk 4]
```diff
@@ -458,16 +461,30 @@ private static Base64URL encodeD(EdECPrivateKey edKey) {
      * Encodes the public key part of an EdDSA key as {@link Base64URL}
      */
     @NotNull
-    private static Base64URL encodeX(EdECPoint point) {
-        var bytes = reverseArray(point.getY().toByteArray());
```
Comment (ndr-brt) [inline core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java:462]: now `reverseArray` method is unused, please remove it

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java [hunk 5]
```diff
@@ -458,16 +461,30 @@ private static Base64URL encodeD(EdECPrivateKey edKey) {
...
+    private static Base64URL encodeX(PublicKey publicKey) {
+        var encoded = publicKey.getEncoded();
 
-        // when the X-coordinate of the curve is odd, we flip the highest-order bit of the first (or last, since we reversed) byte
-        if (point.isXOdd()) {
-            var mask = (byte) 128; // is 1000 0000 binary
-            bytes[bytes.length - 1] ^= mask; // XOR means toggle the left-most bit
+        // safety check: key encoding might fail if the key does not support encoding.
```
Comment (ndr-brt) [inline core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: nit: superfluous comment

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java [hunk 6]
```diff
@@ -458,16 +461,30 @@ private static Base64URL encodeD(EdECPrivateKey edKey) {
...
+        if (encoded == null) {
+            throw new IllegalArgumentException("Public key encoding failed.");
         }
 
-        return Base64URL.encode(bytes);
+        try {
+            // parse the ASN.1 structure safely
+            var spki = SubjectPublicKeyInfo.getInstance(ASN1Primitive.fromByteArray(encoded));
+            // BouncyCastle returns null if the bytes parse to an End-of-Content tag or empty object
```
Comment (ndr-brt) [inline core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: nit: superfluous comment (to what's already mentioned in the exception message)

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java [hunk 7]
```diff
@@ -458,16 +461,30 @@ private static Base64URL encodeD(EdECPrivateKey edKey) {
...
+            if (spki == null) {
+                throw new IllegalArgumentException("Failed to parse the public key ASN.1 structure. The key bytes may be corrupted.");
+            }
+
+            var bytes = spki.getPublicKeyData().getBytes();
+
+            return Base64URL.encode(bytes);
+
+        } catch (IOException e) {
+            // wrap the checked exception in a standard unchecked exception
```
Comment (ndr-brt) [inline core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: nit: superfluous comment

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 8]
```diff
@@ -459,4 +467,120 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
         }
     }
 
+    @Test
```
Comment (ndr-brt) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: Since this test class is getting bigger, could you please group the tests related to `createJwt` in a nested dedicated subclass? That will make the tests easier to browse

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 9]
```diff
@@ -459,4 +467,120 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
...
+    void createJwk_withValidEd25519Key_correctlyEncodesXcoordinate() throws Exception {
+        // generate a Ed25519 key pair and create JWK
```
Comment (ndr-brt) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: nit: please remove all the superfluous comments from the added tests like this one

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 10]
```diff
@@ -459,4 +467,120 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
...
+        assertThatThrownBy(() -> CryptoConverter.createJwk(badKeyPair))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("Failed to parse the public key ASN.1 structure")
+                .hasCauseInstanceOf(IOException.class);
+    }
+
+    @Test
+    void createJwk_withLeadingZeroInXcoordinate_preservesAll32Bytes() throws Exception {
+        // generating an array of exactly 31 bytes to simulate BigInteger dropping a leading zero
+        byte[] shortBytes = new byte[31];
```
Comment (ndr-brt) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: nit: please use `var` when possible

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java [hunk 11]
```diff
@@ -458,16 +449,26 @@ private static Base64URL encodeD(EdECPrivateKey edKey) {
...
-            var mask = (byte) 128; // is 1000 0000 binary
-            bytes[bytes.length - 1] ^= mask; // XOR means toggle the left-most bit
+        if (encoded == null) {
+            throw new IllegalArgumentException("Public key encoding failed.");
         }
 
-        return Base64URL.encode(bytes);
+        try {
+            var spki = SubjectPublicKeyInfo.getInstance(ASN1Primitive.fromByteArray(encoded));
+            if (spki == null) {
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: this null-check is redundant: `.getInstance()` only returns `null` if passed `null`. This can be evoided if the check above is extended to `encoded == null || encoded.length == 0`. This will also get rid of nested try-catch-throw structures.

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 12]
```diff
@@ -459,4 +469,112 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
...
+
+            var mockPubKey = createMockEd25519PublicKey(shortBytes);
+            var badKeyPair = new KeyPair(mockPubKey, null);
+
+            assertNotNull(badKeyPair, "Could not generate a key with a leading zero for the test");
+
+            assertThatThrownBy(() -> CryptoConverter.createJwk(badKeyPair))
+                    .isInstanceOf(IllegalArgumentException.class)
+                    .hasMessage("Failed to parse the public key ASN.1 structure. The key bytes may be corrupted.")
+                    .hasNoCause();
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: better to use `.hasRootCauseInstanceOf`

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 13]
```diff
@@ -459,4 +469,112 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
...
+            var badKeyPair = new KeyPair(mockCorruptedKey, null);
+
+            assertThatThrownBy(() -> CryptoConverter.createJwk(badKeyPair))
+                    .isInstanceOf(IllegalArgumentException.class)
+                    .hasMessageContaining("Failed to parse the public key ASN.1 structure")
+                    .hasCauseInstanceOf(IOException.class);
+        }
+
+        @Test
+        void createJwk_withLeadingZeroInXcoordinate_preservesAll32Bytes() throws Exception {
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: exception never thrown, pls clean up

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 14]
```diff
@@ -459,4 +469,112 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
...
+            assertNotNull(badKeyPair, "Could not generate a key with a leading zero for the test");
+
+            assertThatThrownBy(() -> CryptoConverter.createJwk(badKeyPair))
+                    .isInstanceOf(IllegalArgumentException.class)
+                    .hasMessage("Failed to parse the public key ASN.1 structure. The key bytes may be corrupted.")
+                    .hasNoCause();
+        }
+
+        @Test
+        void createJwk_edDsaKey_handles33ByteYcoordinate_stripsLeadingZero() throws Exception {
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: exception never thrown, pls clean up

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 15]
```diff
@@ -459,4 +469,112 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
...
+
+        @Test
+        void createJwk_withLeadingZeroInXcoordinate_preservesAll32Bytes() throws Exception {
+            var shortBytes = new byte[31];
+            Arrays.fill(shortBytes, (byte) 0xFF);
+
+            var mockPubKey = createMockEd25519PublicKey(shortBytes);
+            var badKeyPair = new KeyPair(mockPubKey, null);
+
+            assertNotNull(badKeyPair, "Could not generate a key with a leading zero for the test");
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: use assertJ assertions. also in other places.

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java (hunk 11 above)
Comment (nikschul) [reply core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: I removed the check completely, only remaining with a single try-catch block

Comment (paullatzelsperger) [reply core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: no, please don't do that. feeding a byte array, that is potentially null or empty will only obscure the cause when an NPE is raised, which will happen if the byte array is null or empty.
I outlined an approach already, please follow it.

File: core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java [hunk 16]
```diff
@@ -459,4 +464,106 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) th
...
+            // X coordinate should be present and properly encoded (32 bytes = 43 chars in Base64URL)
+            assertThat(okp.getX()).isNotNull();
+            assertThat(okp.getX().toString()).hasSize(43);
+            // D coordinate (private key) should also be present and properly encoded
+            assertThat(okp.getD()).isNotNull();
+            assertThat(okp.getD().toString()).hasSize(43);
+        }
+
+        @Test
+        void encodeX_withFaultyEd25519Key_31Bytes_throwsIllegalArgumentException() throws Exception {
```
Comment (paullatzelsperger) [inline core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: This test and the next one are testing effectively the same thing: verifying that ASN.1 parsing fails on garbage bytes. That is still fine, but in order to do that, both tests could be collapsed into one. There is no difference whether we feed 31 or 33 garbage bytes.

Comment (nikschul) [reply core/common/lib/crypto-common-lib/src/test/java/org/eclipse/edc/security/token/jwt/CryptoConverterTest.java]: tests are updated

File: core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java (hunk 11 above)
Comment (nikschul) [reply core/common/lib/crypto-common-lib/src/main/java/org/eclipse/edc/security/token/jwt/CryptoConverter.java]: I added the test for it again, the second try/catch is necessary for the checked IOException of ASN1Primitive.fromByteArray()
